Use new error type & log in Sub/UnsubProposal

anagha-ks opened the pull request:
<!-- Provide a general summary of your changes in the title above

Please read our contribution guidelines and sign the Contributor License
Agreement (CLA) before submitting the pull request. Also, check if there are no
other open pull requests targeting the same issue. -->

#### Description
For SubPayChProposals, UnsubPayProposals API:
- Update it to return the newly defined error type (#132).

- Log the request, error/success response are logged in a structured
format (#133).

##### Category

<!-- Bug Fix -->
<!-- Improvement -->
Implementation Task

##### Relevant issue

Relates to #132, #133.

#### Testing
<!-- Tell us how you have tested the changes. -->

##### Steps to run the tests
<!-- Describe a set of steps to run the tests relevant to this change. -->

1. Updated `Test_SubUnsubChProposal` in `session` package
2. Updated `Test_SubPayChProposals` and `Test_UnsubPayChProposals` in `payment` package
2. 
3.  

#### Checklist 
<!-- Please check if the pull request fulfils these requirements: -->

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

anagha-ks committed:
Use new error type & log in Sub/UnsubProposal

For SubPayChProposals, UnsubPayProposals API:
- Update it to return the newly defined error type (#132).

- Log the request, error/success response are logged in a structured
format. Include method name in each of the lof entries (#133).

Relates to #132, #133

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>

manoranjith commented on session/session.go:
```diff
@@ -520,17 +522,22 @@ func chProposalNotif(parts []string, curr string, chProposal *pclient.BaseChanne
-func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) error {
-	s.Debug("Received request: session.SubChProposals")
+func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) perun.APIErrorV2 {
+	s.WithField("method", "SubChProposals").Info("Received request: session.SubChProposals")
```
Since the method names is logged as a field, you could ignore it in the log message. Instead you could include the parameters. See line line 147 (AddPeerID) for reference.

manoranjith replied:
I have added a suggestion to this line, so that we use uniform error/info messages across all the APIs.

manoranjith commented on session/session.go:
```diff
@@ -539,23 +546,30 @@ func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) error {
 		go s.chProposalNotifier(s.chProposalNotifsCache[0])
 		s.chProposalNotifsCache = s.chProposalNotifsCache[1:i]
 	}
+	s.WithField("method", "SubChProposals").Info("Subscription done successfully")
```
```suggestion
	s.WithField("method", "SubChProposals").Info("Subscribed successfully")
```

manoranjith commented on session/session.go:
```diff
@@ -539,23 +546,30 @@ func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) error {
-func (s *Session) UnsubChProposals() error {
-	s.Debug("Received request: session.UnsubChProposals")
+func (s *Session) UnsubChProposals() perun.APIErrorV2 {
+	s.WithField("method", "UnsubChProposals").Info("Received request: session.UnsubChProposals")
```
See the [comment](https://github.com/hyperledger-labs/perun-node/pull/153/files#r578310910) above.

anagha-ks replied:
Since there is receiving no parameter, ignoring additional log message.

manoranjith replied:
I have added a suggestion to this line, so that we use uniform error/info messages across all the APIs.

manoranjith commented on session/session_test.go:
```diff
@@ -404,11 +404,21 @@ func Test_SubUnsubChProposal(t *testing.T) {
 	t.Run("Sub_sessionClosed", func(t *testing.T) {
 		err = closedSession.SubChProposals(dummyNotifier)
 		require.Error(t, err)
+		wantMessage := session.ErrSessionClosed.Error()
```
Can you update the the sub-tests 2 and 4 to include assertions on error category, code, message etc.,  ?

manoranjith commented on session/session.go:
```diff
@@ -520,17 +522,22 @@ func chProposalNotif(parts []string, curr string, chProposal *pclient.BaseChanne
 
 	if s.chProposalNotifier != nil {
-		return perun.ErrSubAlreadyExists
+		apiErr = perun.NewAPIErrV2ResourceExists("SubChProposals", s.ID(), ErrSubAlreadyExists.Error())
```
```suggestion
		apiErr = perun.NewAPIErrV2ResourceExists("subscription to channel proposals", s.ID(), ErrSubAlreadyExists.Error())
```

Because the phrase will be included in additional info field of error message.

manoranjith commented on session/session.go:
```diff
@@ -539,23 +546,30 @@ func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) error {
 
 	if s.chProposalNotifier == nil {
-		return perun.ErrNoActiveSub
+		apiErr = perun.NewAPIErrV2ResourceNotFound("proposal id", s.ID(), ErrNoActiveSub.Error())
```
```suggestion
		apiErr = perun.NewAPIErrV2ResourceNotFound("subscription to channel proposals", s.ID(), ErrNoActiveSub.Error())
```

The resource is `subscription to channel proposals` and since there can be only one subscription per session, we use the session ID as the ID for the channel proposal.

manoranjith commented on app/payment/session_test.go:
```diff
@@ -161,16 +161,16 @@ func Test_SubPayChProposals(t *testing.T) {
-		require.NoError(t, gotErr)
-		require.NotNil(t, notifier)
+		assert.NoError(t, gotErr)
+		assert.Nil(t, gotErr)
```
What is the reason for changing `require` to `assert` and removing `NotNil` check on notifier  ?

anagha-ks replied:
Keeping the older format.

anagha-ks committed:
Address review comments

Relates to #132, #133

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>

manoranjith commented on session/session.go:
```diff
@@ -523,7 +523,7 @@ func chProposalNotif(parts []string, curr string, chProposal *pclient.BaseChanne
 // SubChProposals implements sessionAPI.SubChProposals.
 func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) perun.APIErrorV2 {
-	s.WithField("method", "SubChProposals").Info("Received request: session.SubChProposals")
+	s.WithField("method", "SubChProposals").Info("Received request: notifier")
```
s.WithField("method", "SubChProposals").Info("Received request with params:", notifier")

manoranjith commented on session/session.go:
```diff
@@ -546,13 +546,13 @@ func (s *Session) SubChProposals(notifier perun.ChProposalNotifier) perun.APIErr
 // UnsubChProposals implements sessionAPI.UnsubChProposals.
 func (s *Session) UnsubChProposals() perun.APIErrorV2 {
-	s.WithField("method", "UnsubChProposals").Info("Received request: session.UnsubChProposals")
+	s.WithField("method", "UnsubChProposals")
```
s.WithField("method", "UnsubChProposals").Info("Received request")

anagha-ks committed:
Change info message

Change error/info message in two places in session sub/unsub channel proposals

Relates to #132, #133

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>
